cmd/pipe_supported.go

```diff
@@ -0,0 +1,77 @@
+//go:build linux
+
+// Copyright (c) 2015-2023 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"golang.org/x/sys/unix"
+)
+
+const pipeMaxSizeProcFile = "/proc/sys/fs/pipe-max-size"
+
+var unsupportedPlatformErr = errors.New("This feature is supported only on Linux systems")
+
+func setPipeSize(fd uintptr, size int) error {
+	_, err := unix.FcntlInt(fd, unix.F_SETPIPE_SZ, size)
+	return err
+}
+
+func getPipeSize(fd uintptr) (int, error) {
+	currentPipeSize, err := unix.FcntlInt(fd, unix.F_GETPIPE_SZ, 0)
+	if err != nil {
+		return 0, err
+	}
+	return currentPipeSize, nil
+}
+
+func getConfiguredMaxPipeSize() (int, error) {
+	b, err := os.ReadFile(pipeMaxSizeProcFile)
+	if err != nil {
+		return 0, err
+	}
+	maxSize, err := strconv.ParseInt(strings.TrimSpace(string(b)), 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("error parsing %s content: %v\n", pipeMaxSizeProcFile,
+			err)
+	}
+	return int(maxSize), nil
+}
+
+// increasePipeBufferSize attempts to increase the pipe size to the system-max,
+// or the input value.
+func increasePipeBufferSize(f *os.File, desiredPipeSize int) {
+	fd := f.Fd()
+
+	if desiredPipeSize <= 0 {
+		maxSize, err := getConfiguredMaxPipeSize()
+		if err == nil {
+			desiredPipeSize = maxSize
+		}
+	}
+
+	if desiredPipeSize > 0 {
+		setPipeSize(fd, desiredPipeSize)
```

Comment:
When set by user on CLI we probably want to error out if we failed to increase it to the desired value.